src/theory/booleans/proof_checker.cpp

```diff
@@ -304,79 +311,79 @@ Node BoolProofRuleChecker::checkInternal(ProofRule id,
-          }
-          else
-          {
-            Assert(args[j] == falseNode);
-            elim.insert(args[j + 1].notNode());
-          }
-        }
+        lhsElim = lits[i];
+        rhsElim = lits[i].notNode();
       }
-      // literal to be removed from "second" clause. They will be negated
-      if (i > 0)
+      else
       {
-        std::size_t index = 2 * (i - 1) + 1;
-        Node pivot = args[index] == trueNode ? args[index + 1].notNode()
-                                             : args[index + 1];
-        elim.insert(pivot);
+        Assert(pols[i] == falseNode);
+        lhsElim = lits[i].notNode();
+        rhsElim = lits[i];
       }
-      Trace("bool-pfcheck") << i << ": elimination set: " << elim << "\n";
-      // only add to conclusion nodes that are not in elimination set. First get
-      // the nodes.
-      //
-      // Since a Node cannot hold an OR with a single child we need to
-      // disambiguate singleton clauses that are OR nodes from non-singleton
-      // clauses (i.e. unit clauses in the SAT solver).
-      //
-      // If the child is not an OR, it is a singleton clause and we take the
-      // child itself as the clause. Otherwise the child can only be a singleton
-      // clause if the child itself is used as a resolution literal, i.e. if the
-      // child is in lhsElim or is equal to rhsElim (which means that the
-      // negation of the child is in lhsElim).
-      std::vector<Node> lits;
-      if (children[i].getKind() == Kind::OR && !elim.count(children[i]))
+      // The index of the child corresponding to the current rhs clause
+      size_t childIndex = i + 1;
+      // Get rhs clause. It's a singleton if not an OR node or if equal to
+      // rhsElim
+      if (children[childIndex].getKind() != Kind::OR
+          || children[childIndex] == rhsElim)
       {
-        lits.insert(lits.end(), children[i].begin(), children[i].end());
+        rhsClause.push_back(children[childIndex]);
       }
       else
       {
-        lits.push_back(children[i]);
+        rhsClause = {children[childIndex].begin(), children[childIndex].end()};
       }
-      Trace("bool-pfcheck") << i << ": clause lits: " << lits << "\n";
-      std::vector<Node> added;
-      for (std::size_t j = 0, size = lits.size(); j < size; ++j)
+      Trace("bool-pfcheck") << i << "-th res link:\n";
+      Trace("bool-pfcheck") << "\t - lhsClause: " << lhsClause << "\n";
+      Trace("bool-pfcheck") << "\t\t - lhsElim: " << lhsElim << "\n";
+      Trace("bool-pfcheck") << "\t - rhsClause: " << rhsClause << "\n";
+      Trace("bool-pfcheck") << "\t\t - rhsElim: " << rhsElim << "\n";
+      // Compute the resulting clause, which will be the next lhsClause, as
+      // follows:
+      //   - remove all lhsElim from lhsClause
+      //   - remove all rhsElim from rhsClause and add the lits to lhsClause
+      lhsClause.erase(std::remove(lhsClause.begin(), lhsClause.end(), lhsElim),
```

Comment:
I don't follow this line.  You are removing all lhsElim from lhsClause, then doing an erase operation on the result clause?

Reply:
It's this: https://en.wikipedia.org/wiki/Erase%E2%80%93remove_idiom

The `std::remove` call will shuffle the elements different from `lhsElim` to the beginning of the new container, and return an iterator to the "rest" of the container (the elements to be removed). Then `erase` will remove the range from there to the end.

This is really ugly though... in C++ 20 one has a native solution, but for now I guess it suffices to add the above as documentation?

Reply:
I've added this as a comment, PTAL